pkg/operator/bootstrapteardown/waitforceo.go

```diff
@@ -0,0 +1,71 @@
+package bootstrapteardown
+
+import (
+	"context"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	operatorversionedclient "github.com/openshift/client-go/operator/clientset/versioned"
+	operatorv1helpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+	clientwatch "k8s.io/client-go/tools/watch"
+	"k8s.io/klog"
+)
+
+func WaitForEtcdBootstrap(ctx context.Context, config *rest.Config) error {
+	operatorConfigClient, err := operatorversionedclient.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+	if err := waitForEtcdBootstrap(ctx, operatorConfigClient.OperatorV1().RESTClient()); err != nil {
+		klog.Errorf("error watching etcd: %#v", err)
+	}
+	return err
+}
+
+func waitForEtcdBootstrap(ctx context.Context, operatorRestClient rest.Interface) error {
+	_, err := clientwatch.UntilWithSync(
+		ctx,
+		cache.NewListWatchFromClient(operatorRestClient, "etcds", "", fields.OneTermEqualSelector("metadata.name", "cluster")),
+		&operatorv1.Etcd{},
+		nil,
+		func(event watch.Event) (bool, error) {
+			switch event.Type {
+			case watch.Added, watch.Modified:
+				etcd, ok := event.Object.(*operatorv1.Etcd)
+				if !ok {
+					klog.Warningf("Expected an Etcd object but got a %q object instead", event.Object.GetObjectKind().GroupVersionKind())
+					return false, nil
+				}
+				return done(etcd)
+			}
+			klog.Infof("Still waiting for the cluster-etcd-operator to bootstrap...")
+			return false, nil
+		},
+	)
+
+	if err == nil {
```

Comment:
I am not that familiar with `UntilWithSync` but what case will we actually error != nil?

Reply:
One case that comes to my mind is if the context passed is cancelled, it might lead to a non-nil error. Also, I don't see any reason for this to not follow the usual convention of Go in the if blocks being `if err != nil {..}` I will address this. Thanks